File: frontend/Concord/src/api/moderationService.js
import { request } from './httpClient';
import { buildPagingQuery } from './pagination';

export async function getBans(serverId, page, pageSize) {
  return request(`Servers/${serverId}/Bans`, { query: buildPagingQuery(page, pageSize) });
}

export async function banMember(serverId, userId, { Reason = null, ExpiresAtUtc = null } = {}) {
  return request(`Servers/${serverId}/Bans/${userId}`, { method: 'PUT', body: { Reason, ExpiresAtUtc } });
}

export async function unbanUser(serverId, userId) {
  return request(`Servers/${serverId}/Bans/${userId}`, { method: 'DELETE' });
}

export async function getMemberModeration(serverId, userId) {
  return request(`Servers/${serverId}/Members/${userId}/Moderation`);
}

export async function muteMember(serverId, userId) {
  return request(`Servers/${serverId}/Members/${userId}/Mute`, { method: 'PUT' });
}

export async function unmuteMember(serverId, userId) {
  return request(`Servers/${serverId}/Members/${userId}/Mute`, { method: 'DELETE' });
}

export async function timeoutMember(serverId, userId, { DurationMinutes, Reason = null }) {
  return request(`Servers/${serverId}/Members/${userId}/Timeout`, {
    method: 'PUT',
    body: { DurationMinutes, Reason },
  });
}

export async function removeTimeout(serverId, userId) {
  return request(`Servers/${serverId}/Members/${userId}/Timeout`, { method: 'DELETE' });
}

File: mobile/concord/lib/api/moderation_service.dart
import 'api_client.dart';
import 'models/member_moderation_response.dart';

/// Ban, mute, and timeout (P1) — mirrors `ModerationController`
/// (`Api/V1.0/Servers/{serverId}`). Kick is *not* here: it stays on
/// `ServersService.kickMember` (`DELETE Servers/{id}/Members/{userId}`), matching how the web
/// app's `moderationService.js` also leaves kick on `serversService`.
class ModerationService {
  ModerationService(this._client);

  final ApiClient _client;

  Future<MemberModerationResponse> getMemberModeration(String serverId, String userId) async {
    final data = await _client.get('/Servers/$serverId/Members/$userId/Moderation');
    return MemberModerationResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<MemberModerationResponse> muteMember(String serverId, String userId) async {
    final data = await _client.put('/Servers/$serverId/Members/$userId/Mute');
    return MemberModerationResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<MemberModerationResponse> unmuteMember(String serverId, String userId) async {
    final data = await _client.delete('/Servers/$serverId/Members/$userId/Mute');
    return MemberModerationResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<MemberModerationResponse> timeoutMember(
    String serverId,
    String userId, {
    required int durationMinutes,
    String? reason,
  }) async {
    final data = await _client.put(
      '/Servers/$serverId/Members/$userId/Timeout',
      body: {'DurationMinutes': durationMinutes, 'Reason': reason},
    );
    return MemberModerationResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<MemberModerationResponse> removeTimeout(String serverId, String userId) async {
    final data = await _client.delete('/Servers/$serverId/Members/$userId/Timeout');
    return MemberModerationResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<void> banMember(String serverId, String userId, {String? reason, DateTime? expiresAtUtc}) {
    return _client.put(
      '/Servers/$serverId/Bans/$userId',
      body: {'Reason': reason, 'ExpiresAtUtc': expiresAtUtc?.toIso8601String()},
    );
  }
}
